### ISSUE TITLE
Operator '<' ***CAN*** be applied to types 'number | BigInt' and 'number'

### ISSUE THREAD

**Issue by coolaj86** [microsoft/TypeScript#57953]
### 🔎 Search Terms

"Operator '<' cannot be applied to types 'number | BigInt' and 'number'"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (v5.3 and latest v5.4)

### 💻 Code

Small demonstration:

```ts
/** 
  * @param {Number|BigInt} n
  */
function isLessThanFive(n) {
  return n < 5;
}
```

My actual code:

```ts
 /**
   * @param {BigInt|Number} n
   * @returns {Number}
   */
  TxUtils.toVarIntSize = function (n) {
    if (n < 253) {
      return 1;
    }

    if (n <= MAX_U16) {
      return 3;
    }

    if (n <= MAX_U32) {
      return 5;
    }

    if (n <= MAX_U64) {
      return 9;
    }

    let err = new Error(E_TOO_BIG_INT);
    Object.assign(err, { code: "E_TOO_BIG_INT" });
    throw err;
  };
```

### 🙁 Actual behavior

> Operator '<' cannot be applied to types 'number | BigInt' and 'number'

### 🙂 Expected behavior

There should not be artificial constraints on Number / BigInt comparison.

If it works in JavaScript, `tsc` should not complain.

**Comment by fatcerberus** [comment]
> If it works in JavaScript, tsc should not complain.

Obligatory link to https://stackoverflow.com/questions/41750390/what-does-all-legal-javascript-is-legal-typescript-mean

**Comment by MartinJohns** [comment]
> If it works in JavaScript, `tsc` should not complain.

I mean.. the entire concept of TypeScript is to prevent things that work in JavaScript.

**Comment by nmain** [comment]
> https://stackoverflow.com/questions/41750390/what-does-all-legal-javascript-is-legal-typescript-mean

How is that relevant in this specific case?  As far as I know, `(number | bigint) > (number | bigint)` has no pitfalls:  There are no situations where two inputs fail to compare the way one would expect them to.  There also isn't a more explicit way to express this comparison.

**Comment by RyanCavanaugh** [comment]
Normally this is where I say "Ah actually there's a hidden gotcha here" but there really isn't one. JS actually does this right in every case.

**Comment by coolaj86** [comment]
@RyanCavanaugh where's the approximate place in the code where this would need to be changed? Is it something that someone unfamiliar with the codebase is likely to be able to find?

(I mean, it _seems_ simple - find some operator list and add a type to a list of comparable types)

**Comment by fatcerberus** [comment]
@nmain The point of linking that was simply to point out that, *in general*, "this code works in JS" is not a motivating argument against TS type errors by itself--it doesn't move the needle at all.  You need more justification than that.

That being said, I'm *very* surprised to find out this comparison is allowed, given that from what I recall the spec authors were very careful to ensure mixing bigints and numbers was a runtime error in most other situations.  `+5n` is a runtime error, for example.

**Comment by Zzzen** [comment]
The real issue here is that `BigInt` does not refer to `bigint` as `Number` does in the JSDoc type expression. It works fine if you replace it with `bigint`.

```ts
/** 
  * @param {Number|bigint} n
  */
function isLessThanFive(n) {
  return n < 5;
}
```

https://www.typescriptlang.org/play?useDefineForClassFields=true&target=9&module=1&noImplicitUseStrict=true&ts=5.2.0-beta&filetype=js#code/PQKhAICh3CAEAOBDATkgtuA3gOQK7oBGApigD6ECWA5pQHYAuAvuHdLMJAGZ50DGDSgHs64SgGcAMsXHiAKgAskdAGKUAbsQAUdAJTZ2KYgzwpRogDzgArAG5ITIA

**Comment by RolandLocke** [comment]
I've been looking at this issue:

`bigint` and `number` primitives can be compared directly using relational operators. Object types (`Number` and `BigInt`) should be coerced to their respective primitive types for comparisons, as JavaScript does implicitly. However, TypeScript's type system disallows comparisons involving `Number` and `BigInt`, even though JavaScript handles these via implicit coercion. 

For example:

`number < bigint` works.
`Number < bigint` or `number < BigInt` and `Number < BigInt` throws a type error in TypeScript.

[Section 7.2.13 of ECMAScript documenting the comparison operations.](https://262.ecma-international.org/15.0/index.html?_gl=1*1v4ijpv*_ga*MTA5Mjc4MTUwMC4xNzMxOTQ3MTgx*_ga_TDCK4DWEPP*MTczMTk0NzE4MC4xLjEuMTczMTk0NzE5MC4wLjAuMA..#sec-islessthan) 

I am posing the following questions:

Should TypeScript align with JavaScript behavior by allowing comparisons between `Number/BigInt` objects and their primitive counterparts or only between `Number` and `BigInt` object types?

Any insights would be greatly appreciated! Thank you!
